### DISCUSSION TITLE
Discussion on Chat Memory components

### DISCUSSION THREAD

**Discussion by TuanaCelik** [deepset-ai/haystack-experimental#75] [General]
This is the discussion board for a group of experimental components introduced as a solution to implement chat memory for chat with RAG-like use-cases.
- [ChatMessageWriter](https://github.com/deepset-ai/haystack-experimental/blob/main/haystack_experimental/components/writers/chat_message_writer.py)
- [ChatMessageRetriever](https://github.com/deepset-ai/haystack-experimental/blob/main/haystack_experimental/components/retrievers/chat_message_retriever.py)
- [InMemoryChatMessageStore](https://github.com/deepset-ai/haystack-experimental/blob/main/haystack_experimental/chat_message_stores/in_memory.py)

🧑‍🍳 **[Try the Cookbook here](https://colab.research.google.com/github/deepset-ai/haystack-cookbook/blob/main/notebooks/conversational_rag_using_memory.ipynb)**

Let us know how you get along with these components, and please do let us know if you have any feedback.

> **Note:** Experimental features live in this repository for a fixed period of time. We don't guarantee that we will continue maintaining experimental features. But if they are successful, stable and if you like it, we will move the feature to the core Haystack package.

PS: The contributor for the first version of these experimental components is @vblagoje 🚀

**Comment by shadeMe** [comment]
From @anakin87 

While preparing a notebook on Tools, I needed some form of short-term memory of the conversation.
I tried to use `ChatMessageStore`, `ChatMessageWriter`, and`ChatMessageRetriever`.

Unfortunately, I have encountered some **roadblocks with the `ChatMessageRetriever`**.
You can find my experiments [in this notebook](https://colab.research.google.com/drive/1PKVO3Gemacl_7EcDRGT0O1i9Cq5tczUF?usp=sharing).

---

Main problems I have encountered:
`ChatMessageRetriever` has no input sockets
- cannot receive any input from other components
- cannot be placed in a conditional branch
- must be placed at the beginning of a Pipeline
- since it does not need inputs, it always runs

I am aware of the ongoing work on the ChatMessageRetriever (see deepset-ai/haystack#8258), but my impression is that most of these issues will remain.

 Given that my use case is not very exotic, I think it would be beneficial to reflect on how to improve and make this experimental component more flexible.

**Comment by kkmarv** [comment]
I want to leave my feedback as well and say that I greatly appreciate the Chat Memory components and think they are very useful. In my case I've used them for bootstrapping small examples I can show to colleagues.

However, I just want to point out a small inconsistency in the cookbook. In [Build The Pipeline](https://haystack.deepset.ai/cookbook/conversational_rag_using_memory#build-the-pipeline) it says that a `ListJoiner` component will come to Haystack in version 2.8.0 which has just released. This is confusing since version 2.8.0 did not implement such a component. I would be glad to use it though!

**Comment by lambda-science** [comment]
About the `InMemoryChatMessageStore`
- To solve multiple call leading to confusion of conversation: the `self.messages` should probably be a dictionnary with a `ThreadID/ChatID` as key and list of ChatMessage as value. It could be auto-generated if the list of ChatMessage was empty or given by the user.

We should also have a way when running the pipeline to inject ChatMessages in the `InMemoryChatMessageStore`.
For example I build an API based on Haystack, my /chat endpoint can take a `query` and a `chat_history` as parameter that is a list of ChatMessages. This is done to let the developper adjust himself how many messages from history he wants to use instead of letting the pipeline define a `top_k`. This also allow to modify the history or start again a conversation even after the `InMemoryChatMessageStore` has been cleared.
To be more clear: not only the pipeline should be able to create and manage history. We should have input surface to put our own history of messages.

**Comment by lambda-science** [comment]
Also I believe that in CookBooks we shouldn't showcase the usage of Memory/History by feeding the history in a single big prompt. The proper way is to feed a list of ChatMessage in chronological order directly to the ChatGenerator.

**Comment by mariosconsta** [comment]
I am attaching my feature request here so the devs can keep track of it: [Session Based chat message store](https://github.com/deepset-ai/haystack/discussions/9328).

I would like to make an addition as well other than the session idea:

-- Being able to load a conversation through a dictionary. Open WebUI's pipelines provide the conversation in the form of a dictionary. So if you want to resume a conversation that is not saved on memory, you can do so by using the dictionary.

**Comment by LokeshD-off** [comment]
I think we all are missing out on the most important feature to be required in here which is to incorporate multiple user's conversation history and not get them mixed up

**Reply by bilgeyucel** [reply]
Hi @LokeshD-off, that's a good point. I think, this can be implemented in a very similar way to metadata filtering

**Comment by svenstehle** [comment]
Hi all, if this is not the correct place to ask this question or raise this issue, please point me to the correct place and I am happy to move there :)

I am trying to set up a ValidationLoop according to this [Tutorial: Generating Structured Output with Loop-Based Auto-Correction](https://haystack.deepset.ai/tutorials/28_structured_output_with_loop) in my pipeline. This tutorial works, the Loop works for me.

However adding memory interferes with the Looping. I have added memory to the ValidationLoop example in a similar way as in my pipeline. Basically from this example: [Tutorial: Creating a Multi-Agent System with Haystack](https://haystack.deepset.ai/cookbook/conversational_rag_using_memory)

Expected Behavior: Loop continues until successful result or pipeline exits after max retries.

Actual behavior: Loop does not continue, instead the pipeline jumps to the memory_joiner and memory_writer and skips all the other steps which leads to code failure. 

- For the minimal example, when I remove "memories" from required variables in the PromptBuilder, it works. 
- Looping then functions correctly.

However in my pipeline that is not the case. The prompt that is connected to the Looping does not even take in memories! I made memories in my whole code non-required and this did not impact behavior. The pipeline jumps to memory_joiner & memory_writer in the same way as with the Example pipeline when I have "memories" as required variable in the PromptBuilder.

I really want this validation loop to work out together with memory. I need both.

Can someone please help me? I can provide more information and code if required :)
I also have flowcharts of the pipelines.

Thanks,
Sven

**Comment by sjrl** [comment]
Thanks for all the feedback! We have made major updates to these components and started a new discussion thread on them here https://github.com/deepset-ai/haystack-experimental/discussions/420

We have also included a new tutorial on how to use them (linked in the new discussion) and [here](https://haystack.deepset.ai/tutorials/48_conversational_rag)
